src/poly1305.rs

```diff
@@ -87,27 +142,69 @@ pub type Tag = [u8; TAG_LEN];
 /// The length of a `Tag`.
 pub const TAG_LEN: usize = 128 / 8;
 
-type State = [u8; STATE_LEN];
-const STATE_LEN: usize = 256;
+const BLOCK_STATE_SIZE: usize = 192;
+
+#[repr(C)]
+struct Funcs {
+    blocks_fn: unsafe extern fn(*mut c_void, *const u8, c::size_t, u32), 
+    emit_fn: unsafe extern fn(*mut c_void, *mut u8, *const u32),
+}
+
+fn init(state: &mut [u8; BLOCK_STATE_SIZE], key: &[u8; KEY_LEN], func: *mut Funcs) -> i32 {
+    unsafe {
+        GFp_poly1305_init_asm(
+            state.as_mut_ptr() as *mut c_void,
+            key.as_ptr(),
+            func as *mut c_void
+        )
+    }
+}
+
+impl Funcs {
+    fn blocks(&self, state: &mut [u8; BLOCK_STATE_SIZE], data: &[u8], should_pad: u32) {
+        unsafe {
+            (self.blocks_fn)(
+                state.as_mut_ptr() as *mut c_void,
+                data.as_ptr(),
+                data.len(),
+                should_pad
+            )
+        };
+    }
+
+    fn emit(&self, state: &mut [u8; BLOCK_STATE_SIZE], tag: &mut Tag, nonce: &[u32; 4]) {
+        unsafe {
+             (self.emit_fn)(
+                 state.as_mut_ptr() as *mut c_void,
+                 tag.as_mut_ptr(),
+                 nonce.as_ptr()
+             );
+        }
+    }
+}
+
+#[repr(C)]
```

Comment:
I don't think this actually needs to be `#[repr(C)]`, does it? It seems like maybe only `Funcs` needs to be `#[repr(C)`]? If SigningContext doesn't need to be `#[repr(C)]` then you can use more native Rust types, such as `usize` instead of `c::int`, and in particular you can get rid of most/all casts in the code.